protocols/imap/src/main/java/org/apache/james/imap/processor/base/UidMsnMapper.java

```diff
@@ -35,54 +37,101 @@
     public final static int FIRST_MSN = 1;
 
     private final HashBiMap<Integer, MessageUid> msnToUid;
+    private final ReadWriteLock readWriteLock;
 
     public UidMsnMapper() {
+        readWriteLock = new ReentrantReadWriteLock();
```

Comment:
Wow, I just discovered ReadWriteLock thank you so much ... i'm joking, of course. It's not the answer I expected.
So, you say you gain safety on read compared to synchronized ? I don't agree. You may gain in speed but probably don't worth it : this class is probably not used in concurrent situation to justify your complex approach.
WDYT ?